File: go/core/colombia-public_sdk.go
package core

import (
	"fmt"

	vs "github.com/voxgig-sdk/colombia-public-sdk/go/utility/struct"
)

type ColombiaPublicSDK struct {
	Mode     string
	options  map[string]any
	utility  *Utility
	Features []Feature
	rootctx  *Context
}

func NewColombiaPublicSDK(options map[string]any) *ColombiaPublicSDK {
	sdk := &ColombiaPublicSDK{
		Mode:     "live",
		Features: []Feature{},
	}

	sdk.utility = NewUtility()

	config := MakeConfig()

	sdk.rootctx = sdk.utility.MakeContext(map[string]any{
		"client":  sdk,
		"utility": sdk.utility,
		"config":  config,
		"options": options,
		"shared":  map[string]any{},
	}, nil)

	sdk.options = sdk.utility.MakeOptions(sdk.rootctx)

	if vs.GetPath([]any{"feature", "test", "active"}, sdk.options) == true {
		sdk.Mode = "test"
	}

	sdk.rootctx.Options = sdk.options

	// Add features from config.
	featureOpts := ToMapAny(vs.GetProp(sdk.options, "feature"))
	if featureOpts != nil {
		for _, item := range vs.Items(featureOpts) {
			fname, _ := item[0].(string)
			fopts := ToMapAny(item[1])
			if fopts != nil {
				if active, ok := fopts["active"]; ok {
					if ab, ok := active.(bool); ok && ab {
						sdk.utility.FeatureAdd(sdk.rootctx, makeFeature(fname))
					}
				}
			}
		}
	}

	// Add extension features.
	if extend := vs.GetProp(sdk.options, "extend"); extend != nil {
		if extList, ok := extend.([]any); ok {
			for _, f := range extList {
				if feat, ok := f.(Feature); ok {
					sdk.utility.FeatureAdd(sdk.rootctx, feat)
				}
			}
		}
	}

	// Initialize features.
	for _, f := range sdk.Features {
		sdk.utility.FeatureInit(sdk.rootctx, f)
	}

	sdk.utility.FeatureHook(sdk.rootctx, "PostConstruct")

	return sdk
}

func (sdk *ColombiaPublicSDK) OptionsMap() map[string]any {
	out := vs.Clone(sdk.options)
	if om, ok := out.(map[string]any); ok {
		return om
	}
	return map[string]any{}
}

func (sdk *ColombiaPublicSDK) GetUtility() *Utility {
	return CopyUtility(sdk.utility)
}

func (sdk *ColombiaPublicSDK) GetRootCtx() *Context {
	return sdk.rootctx
}

func (sdk *ColombiaPublicSDK) Prepare(fetchargs map[string]any) (map[string]any, error) {
	utility := sdk.utility

	if fetchargs == nil {
		fetchargs = map[string]any{}
	}

	var ctrl map[string]any
	if c := vs.GetProp(fetchargs, "ctrl"); c != nil {
		if cm, ok := c.(map[string]any); ok {
			ctrl = cm
		}
	}
	if ctrl == nil {
		ctrl = map[string]any{}
	}

	ctx := utility.MakeContext(map[string]any{
		"opname": "prepare",
		"ctrl":   ctrl,
	}, sdk.rootctx)

	options := sdk.options

	path, _ := vs.GetProp(fetchargs, "path").(string)
	method, _ := vs.GetProp(fetchargs, "method").(string)
	if method == "" {
		method = "GET"
	}

	params := ToMapAny(vs.GetProp(fetchargs, "params"))
	if params == nil {
		params = map[string]any{}
	}
	query := ToMapAny(vs.GetProp(fetchargs, "query"))
	if query == nil {
		query = map[string]any{}
	}

	headers := utility.PrepareHeaders(ctx)

	base, _ := vs.GetProp(options, "base").(string)
	prefix, _ := vs.GetProp(options, "prefix").(string)
	suffix, _ := vs.GetProp(options, "suffix").(string)

	ctx.Spec = NewSpec(map[string]any{
		"base":    base,
		"prefix":  prefix,
		"suffix":  suffix,
		"path":    path,
		"method":  method,
		"params":  params,
		"query":   query,
		"headers": headers,
		"body":    vs.GetProp(fetchargs, "body"),
		"step":    "start",
	})

	// Merge user-provided headers.
	if uh := vs.GetProp(fetchargs, "headers"); uh != nil {
		if uhm, ok := uh.(map[string]any); ok {
			for k, v := range uhm {
				ctx.Spec.Headers[k] = v
			}
		}
	}

	_, err := utility.PrepareAuth(ctx)
	if err != nil {
		return nil, err
	}

	return utility.MakeFetchDef(ctx)
}

func (sdk *ColombiaPublicSDK) Direct(fetchargs map[string]any) (map[string]any, error) {
	utility := sdk.utility

	fetchdef, err := sdk.Prepare(fetchargs)
	if err != nil {
		return map[string]any{"ok": false, "err": err}, nil
	}

	if fetchargs == nil {
		fetchargs = map[string]any{}
	}

	var ctrl map[string]any
	if c := vs.GetProp(fetchargs, "ctrl"); c != nil {
		if cm, ok := c.(map[string]any); ok {
			ctrl = cm
		}
	}
	if ctrl == nil {
		ctrl = map[string]any{}
	}

	ctx := utility.MakeContext(map[string]any{
		"opname": "direct",
		"ctrl":   ctrl,
	}, sdk.rootctx)

	url, _ := fetchdef["url"].(string)
	fetched, fetchErr := utility.Fetcher(ctx, url, fetchdef)

	if fetchErr != nil {
		return map[string]any{"ok": false, "err": fetchErr}, nil
	}

	if fetched == nil {
		return map[string]any{
			"ok":  false,
			"err": ctx.MakeError("direct_no_response", "response: undefined"),
		}, nil
	}

	if fm, ok := fetched.(map[string]any); ok {
		status := ToInt(vs.GetProp(fm, "status"))
		headers := vs.GetProp(fm, "headers")

		// No-body responses (204, 304) and explicit zero content-length
		// must skip JSON parsing — calling json() on an empty body errors.
		var contentLength string
		if hm, ok := headers.(map[string]any); ok {
			if cl, ok := hm["content-length"]; ok {
				contentLength = fmt.Sprintf("%v", cl)
			}
		}
		noBody := status == 204 || status == 304 || contentLength == "0"

		var jsonData any
		if !noBody {
			if jf := vs.GetProp(fm, "json"); jf != nil {
				if f, ok := jf.(func() any); ok {
					// f() returns nil on parse error in our fetcher.
					jsonData = f()
				}
			}
		}

		return map[string]any{
			"ok":      status >= 200 && status < 300,
			"status":  status,
			"headers": headers,
			"data":    jsonData,
		}, nil
	}

	return map[string]any{"ok": false, "err": ctx.MakeError("direct_invalid", "invalid response type")}, nil
}


// Airport returns a Airport entity bound to this client.
// Idiomatic usage: client.Airport(nil).List(nil, nil) or
// client.Airport(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *ColombiaPublicSDK) Airport(data map[string]any) ColombiaPublicEntity {
	return NewAirportEntityFunc(sdk, data)
}


// CategoryNaturalArea returns a CategoryNaturalArea entity bound to this client.
// Idiomatic usage: client.CategoryNaturalArea(nil).List(nil, nil) or
// client.CategoryNaturalArea(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *ColombiaPublicSDK) CategoryNaturalArea(data map[string]any) ColombiaPublicEntity {
	return NewCategoryNaturalAreaEntityFunc(sdk, data)
}


// ConstitutionArticle returns a ConstitutionArticle entity bound to this client.
// Idiomatic usage: client.ConstitutionArticle(nil).List(nil, nil) or
// client.ConstitutionArticle(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *ColombiaPublicSDK) ConstitutionArticle(data map[string]any) ColombiaPublicEntity {
	return NewConstitutionArticleEntityFunc(sdk, data)
}


// Country returns a Country entity bound to this client.
// Idiomatic usage: client.Country(nil).List(nil, nil) or
// client.Country(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *ColombiaPublicSDK) Country(data map[string]any) ColombiaPublicEntity {
	return NewCountryEntityFunc(sdk, data)
}


// Department returns a Department entity bound to this client.
// Idiomatic usage: client.Department(nil).List(nil, nil) or
// client.Department(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *ColombiaPublicSDK) Department(data map[string]any) ColombiaPublicEntity {
	return NewDepartmentEntityFunc(sdk, data)
}


// Holiday returns a Holiday entity bound to this client.
// Idiomatic usage: client.Holiday(nil).List(nil, nil) or
// client.Holiday(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *ColombiaPublicSDK) Holiday(data map[string]any) ColombiaPublicEntity {
	return NewHolidayEntityFunc(sdk, data)
}


// InvasiveSpecie returns a InvasiveSpecie entity bound to this client.
// Idiomatic usage: client.InvasiveSpecie(nil).List(nil, nil) or
// client.InvasiveSpecie(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *ColombiaPublicSDK) InvasiveSpecie(data map[string]any) ColombiaPublicEntity {
	return NewInvasiveSpecieEntityFunc(sdk, data)
}


// Map returns a Map entity bound to this client.
// Idiomatic usage: client.Map(nil).List(nil, nil) or
// client.Map(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *ColombiaPublicSDK) Map(data map[string]any) ColombiaPublicEntity {
	return NewMapEntityFunc(sdk, data)
}


// NativeCommunity returns a NativeCommunity entity bound to this client.
// Idiomatic usage: client.NativeCommunity(nil).List(nil, nil) or
// client.NativeCommunity(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *ColombiaPublicSDK) NativeCommunity(data map[string]any) ColombiaPublicEntity {
	return NewNativeCommunityEntityFunc(sdk, data)
}


// NaturalArea returns a NaturalArea entity bound to this client.
// Idiomatic usage: client.NaturalArea(nil).List(nil, nil) or
// client.NaturalArea(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *ColombiaPublicSDK) NaturalArea(data map[string]any) ColombiaPublicEntity {
	return NewNaturalAreaEntityFunc(sdk, data)
}


// President returns a President entity bound to this client.
// Idiomatic usage: client.President(nil).List(nil, nil) or
// client.President(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *ColombiaPublicSDK) President(data map[string]any) ColombiaPublicEntity {
	return NewPresidentEntityFunc(sdk, data)
}


// Radio returns a Radio entity bound to this client.
// Idiomatic usage: client.Radio(nil).List(nil, nil) or
// client.Radio(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *ColombiaPublicSDK) Radio(data map[string]any) ColombiaPublicEntity {
	return NewRadioEntityFunc(sdk, data)
}


// Region returns a Region entity bound to this client.
// Idiomatic usage: client.Region(nil).List(nil, nil) or
// client.Region(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *ColombiaPublicSDK) Region(data map[string]any) ColombiaPublicEntity {
	return NewRegionEntityFunc(sdk, data)
}


// TouristicAttraction returns a TouristicAttraction entity bound to this client.
// Idiomatic usage: client.TouristicAttraction(nil).List(nil, nil) or
// client.TouristicAttraction(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *ColombiaPublicSDK) TouristicAttraction(data map[string]any) ColombiaPublicEntity {
	return NewTouristicAttractionEntityFunc(sdk, data)
}


// TypicalDish returns a TypicalDish entity bound to this client.
// Idiomatic usage: client.TypicalDish(nil).List(nil, nil) or
// client.TypicalDish(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *ColombiaPublicSDK) TypicalDish(data map[string]any) ColombiaPublicEntity {
	return NewTypicalDishEntityFunc(sdk, data)
}



func TestSDK(testopts map[string]any, sdkopts map[string]any) *ColombiaPublicSDK {
	if sdkopts == nil {
		sdkopts = map[string]any{}
	}
	sdkopts = vs.Clone(sdkopts).(map[string]any)

	if testopts == nil {
		testopts = map[string]any{}
	}
	testopts = vs.Clone(testopts).(map[string]any)
	testopts["active"] = true

	vs.SetPath(sdkopts, []any{"feature", "test"}, testopts)

	sdk := NewColombiaPublicSDK(sdkopts)
	sdk.Mode = "test"

	return sdk
}

File: ts/src/ColombiaPublicSDK.ts
// ColombiaPublic Ts SDK

import { AirportEntity } from './entity/AirportEntity'
import { CategoryNaturalAreaEntity } from './entity/CategoryNaturalAreaEntity'
import { ConstitutionArticleEntity } from './entity/ConstitutionArticleEntity'
import { CountryEntity } from './entity/CountryEntity'
import { DepartmentEntity } from './entity/DepartmentEntity'
import { HolidayEntity } from './entity/HolidayEntity'
import { InvasiveSpecieEntity } from './entity/InvasiveSpecieEntity'
import { MapEntity } from './entity/MapEntity'
import { NativeCommunityEntity } from './entity/NativeCommunityEntity'
import { NaturalAreaEntity } from './entity/NaturalAreaEntity'
import { PresidentEntity } from './entity/PresidentEntity'
import { RadioEntity } from './entity/RadioEntity'
import { RegionEntity } from './entity/RegionEntity'
import { TouristicAttractionEntity } from './entity/TouristicAttractionEntity'
import { TypicalDishEntity } from './entity/TypicalDishEntity'

export type * from './ColombiaPublicTypes'


import { inspect } from 'node:util'

import type { Context, Feature } from './types'

import { config } from './Config'
import { ColombiaPublicEntityBase } from './ColombiaPublicEntityBase'
import { Utility } from './utility/Utility'


import { BaseFeature } from './feature/base/BaseFeature'


const stdutil = new Utility()


class ColombiaPublicSDK {
  _mode: string = 'live'
  _options: any
  _utility = new Utility()
  _features: Feature[]
  _rootctx: Context

  constructor(options?: any) {

    this._rootctx = this._utility.makeContext({
      client: this,
      utility: this._utility,
      config,
      options,
      shared: new WeakMap()
    })

    this._options = this._utility.makeOptions(this._rootctx)

    const struct = this._utility.struct
    const getpath = struct.getpath
    const items = struct.items

    if (true === getpath(this._options.feature, 'test.active')) {
      this._mode = 'test'
    }

    this._rootctx.options = this._options

    this._features = []

    const featureAdd = this._utility.featureAdd
    const featureInit = this._utility.featureInit

    items(this._options.feature, (fitem: [string, any]) => {
      const fname = fitem[0]
      const fopts = fitem[1]
      if (fopts.active) {
        featureAdd(this._rootctx, this._rootctx.config.makeFeature(fname))
      }
    })

    if (null != this._options.extend) {
      for (let f of this._options.extend) {
        featureAdd(this._rootctx, f)
      }
    }

    for (let f of this._features) {
      featureInit(this._rootctx, f)
    }

    const featureHook = this._utility.featureHook
    featureHook(this._rootctx, 'PostConstruct')
  }


  options() {
    return this._utility.struct.clone(this._options)
  }


  utility() {
    return this._utility.struct.clone(this._utility)
  }


  async prepare(fetchargs?: any) {
    const utility = this._utility
    const struct = utility.struct
    const clone = struct.clone

    const {
      makeContext,
      makeFetchDef,
      prepareHeaders,
      prepareAuth,
    } = utility

    fetchargs = fetchargs || {}

    let ctx: Context = makeContext({
      opname: 'prepare',
      ctrl: fetchargs.ctrl || {},
    }, this._rootctx)

    const options = this._options

    // Build spec directly from SDK options + user-provided fetch args.
    const spec: any = {
      base: options.base,
      prefix: options.prefix,
      suffix: options.suffix,
      path: fetchargs.path || '',
      method: fetchargs.method || 'GET',
      params: fetchargs.params || {},
      query: fetchargs.query || {},
      headers: prepareHeaders(ctx),
      body: fetchargs.body,
      step: 'start',
    }

    ctx.spec = spec

    // Merge user-provided headers over SDK defaults.
    if (fetchargs.headers) {
      const uheaders = fetchargs.headers
      for (let key in uheaders) {
        spec.headers[key] = uheaders[key]
      }
    }

    // Apply SDK auth (apikey, auth prefix, etc.)
    const authResult = prepareAuth(ctx)
    if (authResult instanceof Error) {
      return authResult
    }

    return makeFetchDef(ctx)
  }


  async direct(fetchargs?: any) {
    const utility = this._utility
    const fetcher = utility.fetcher
    const makeContext = utility.makeContext

    const fetchdef = await this.prepare(fetchargs)
    if (fetchdef instanceof Error) {
      return fetchdef
    }

    let ctx: Context = makeContext({
      opname: 'direct',
      ctrl: (fetchargs || {}).ctrl || {},
    }, this._rootctx)

    try {
      const fetched = await fetcher(ctx, fetchdef.url, fetchdef)

      if (null == fetched) {
        return { ok: false, err: ctx.error('direct_no_response', 'response: undefined') }
      }
      else if (fetched instanceof Error) {
        return { ok: false, err: fetched }
      }

      const status = fetched.status

      // No body responses (204 No Content, 304 Not Modified) and explicit
      // zero content-length must skip JSON parsing — fetched.json() would
      // throw `Unexpected end of JSON input` on an empty body.
      const headers = fetched.headers
      const contentLength = headers && 'function' === typeof headers.get
        ? headers.get('content-length')
        : (headers || {})['content-length']
      const noBody = 204 === status || 304 === status || '0' === String(contentLength)

      let json: any = undefined
      if (!noBody) {
        try {
          json = 'function' === typeof fetched.json ? await fetched.json() : fetched.json
        }
        catch (parseErr) {
          // Body wasn't valid JSON — surface the raw response rather than
          // throwing. data stays undefined; callers can inspect status/headers.
          json = undefined
        }
      }

      return {
        ok: status >= 200 && status < 300,
        status,
        headers: fetched.headers,
        data: json,
      }
    }
    catch (err: any) {
      return { ok: false, err }
    }
  }



  _airport?: AirportEntity

  // Idiomatic facade: `client.airport.list()` / `client.airport.load({ id })`.
  get airport(): AirportEntity {
    return (this._airport ??= new AirportEntity(this, undefined))
  }

  /** @deprecated Use `client.airport` instead. */
  Airport(data?: any) {
    const self = this
    return new AirportEntity(self,data)
  }


  _category_natural_area?: CategoryNaturalAreaEntity

  // Idiomatic facade: `client.category_natural_area.list()` / `client.category_natural_area.load({ id })`.
  get category_natural_area(): CategoryNaturalAreaEntity {
    return (this._category_natural_area ??= new CategoryNaturalAreaEntity(this, undefined))
  }

  /** @deprecated Use `client.category_natural_area` instead. */
  CategoryNaturalArea(data?: any) {
    const self = this
    return new CategoryNaturalAreaEntity(self,data)
  }


  _constitution_article?: ConstitutionArticleEntity

  // Idiomatic facade: `client.constitution_article.list()` / `client.constitution_article.load({ id })`.
  get constitution_article(): ConstitutionArticleEntity {
    return (this._constitution_article ??= new ConstitutionArticleEntity(this, undefined))
  }

  /** @deprecated Use `client.constitution_article` instead. */
  ConstitutionArticle(data?: any) {
    const self = this
    return new ConstitutionArticleEntity(self,data)
  }


  _country?: CountryEntity

  // Idiomatic facade: `client.country.list()` / `client.country.load({ id })`.
  get country(): CountryEntity {
    return (this._country ??= new CountryEntity(this, undefined))
  }

  /** @deprecated Use `client.country` instead. */
  Country(data?: any) {
    const self = this
    return new CountryEntity(self,data)
  }


  _department?: DepartmentEntity

  // Idiomatic facade: `client.department.list()` / `client.department.load({ id })`.
  get department(): DepartmentEntity {
    return (this._department ??= new DepartmentEntity(this, undefined))
  }

  /** @deprecated Use `client.department` instead. */
  Department(data?: any) {
    const self = this
    return new DepartmentEntity(self,data)
  }


  _holiday?: HolidayEntity

  // Idiomatic facade: `client.holiday.list()` / `client.holiday.load({ id })`.
  get holiday(): HolidayEntity {
    return (this._holiday ??= new HolidayEntity(this, undefined))
  }

  /** @deprecated Use `client.holiday` instead. */
  Holiday(data?: any) {
    const self = this
    return new HolidayEntity(self,data)
  }


  _invasive_specie?: InvasiveSpecieEntity

  // Idiomatic facade: `client.invasive_specie.list()` / `client.invasive_specie.load({ id })`.
  get invasive_specie(): InvasiveSpecieEntity {
    return (this._invasive_specie ??= new InvasiveSpecieEntity(this, undefined))
  }

  /** @deprecated Use `client.invasive_specie` instead. */
  InvasiveSpecie(data?: any) {
    const self = this
    return new InvasiveSpecieEntity(self,data)
  }


  _map?: MapEntity

  // Idiomatic facade: `client.map.list()` / `client.map.load({ id })`.
  get map(): MapEntity {
    return (this._map ??= new MapEntity(this, undefined))
  }

  /** @deprecated Use `client.map` instead. */
  Map(data?: any) {
    const self = this
    return new MapEntity(self,data)
  }


  _native_community?: NativeCommunityEntity

  // Idiomatic facade: `client.native_community.list()` / `client.native_community.load({ id })`.
  get native_community(): NativeCommunityEntity {
    return (this._native_community ??= new NativeCommunityEntity(this, undefined))
  }

  /** @deprecated Use `client.native_community` instead. */
  NativeCommunity(data?: any) {
    const self = this
    return new NativeCommunityEntity(self,data)
  }


  _natural_area?: NaturalAreaEntity

  // Idiomatic facade: `client.natural_area.list()` / `client.natural_area.load({ id })`.
  get natural_area(): NaturalAreaEntity {
    return (this._natural_area ??= new NaturalAreaEntity(this, undefined))
  }

  /** @deprecated Use `client.natural_area` instead. */
  NaturalArea(data?: any) {
    const self = this
    return new NaturalAreaEntity(self,data)
  }


  _president?: PresidentEntity

  // Idiomatic facade: `client.president.list()` / `client.president.load({ id })`.
  get president(): PresidentEntity {
    return (this._president ??= new PresidentEntity(this, undefined))
  }

  /** @deprecated Use `client.president` instead. */
  President(data?: any) {
    const self = this
    return new PresidentEntity(self,data)
  }


  _radio?: RadioEntity

  // Idiomatic facade: `client.radio.list()` / `client.radio.load({ id })`.
  get radio(): RadioEntity {
    return (this._radio ??= new RadioEntity(this, undefined))
  }

  /** @deprecated Use `client.radio` instead. */
  Radio(data?: any) {
    const self = this
    return new RadioEntity(self,data)
  }


  _region?: RegionEntity

  // Idiomatic facade: `client.region.list()` / `client.region.load({ id })`.
  get region(): RegionEntity {
    return (this._region ??= new RegionEntity(this, undefined))
  }

  /** @deprecated Use `client.region` instead. */
  Region(data?: any) {
    const self = this
    return new RegionEntity(self,data)
  }


  _touristic_attraction?: TouristicAttractionEntity

  // Idiomatic facade: `client.touristic_attraction.list()` / `client.touristic_attraction.load({ id })`.
  get touristic_attraction(): TouristicAttractionEntity {
    return (this._touristic_attraction ??= new TouristicAttractionEntity(this, undefined))
  }

  /** @deprecated Use `client.touristic_attraction` instead. */
  TouristicAttraction(data?: any) {
    const self = this
    return new TouristicAttractionEntity(self,data)
  }


  _typical_dish?: TypicalDishEntity

  // Idiomatic facade: `client.typical_dish.list()` / `client.typical_dish.load({ id })`.
  get typical_dish(): TypicalDishEntity {
    return (this._typical_dish ??= new TypicalDishEntity(this, undefined))
  }

  /** @deprecated Use `client.typical_dish` instead. */
  TypicalDish(data?: any) {
    const self = this
    return new TypicalDishEntity(self,data)
  }




  static test(testoptsarg?: any, sdkoptsarg?: any) {
    const struct = stdutil.struct
    const setpath = struct.setpath
    const getdef = struct.getdef
    const clone = struct.clone
    const setprop = struct.setprop

    const sdkopts = getdef(clone(sdkoptsarg), {})
    const testopts = getdef(clone(testoptsarg), {})
    setprop(testopts, 'active', true)
    setpath(sdkopts, 'feature.test', testopts)

    const testsdk = new ColombiaPublicSDK(sdkopts)
    testsdk._mode = 'test'

    return testsdk
  }


  tester(testopts?: any, sdkopts?: any) {
    return ColombiaPublicSDK.test(testopts, sdkopts)
  }


  toJSON() {
    return { name: 'ColombiaPublic' }
  }

  toString() {
    return 'ColombiaPublic ' + this._utility.struct.jsonify(this.toJSON())
  }

  [inspect.custom]() {
    return this.toString()
  }

}




const SDK = ColombiaPublicSDK


export {
  stdutil,

  BaseFeature,
  ColombiaPublicEntityBase,

  ColombiaPublicSDK,
  SDK,
}


